# Proyecto P1 — Análisis exploratorio del Titanic (numpy y pandas)
**Optativa Python para IA — 2º DAM**

Notebook final que integra el trabajo de todo el equipo, organizado por tareas (T1 a T5). Cada sección se desarrolló de forma independiente siguiendo el contrato de datos del README, por eso cada una vuelve a cargar el dataset y aplicar el mismo tratamiento de nulos: así cualquier sección se puede ejecutar o revisar por separado.

**Dataset:** Titanic (versión de seaborn), 891 filas y 15 columnas, en `Data/titanic.csv`.

| Sección | Autor/a |
|---|---|
| T1 Carga e inspección | Jia Xi Shan |
| T2a Nulos | Alejandro |
| T2b Duplicados | Noel (detección) y Daniel (análisis y decisión) |
| T3a Franja de edad | Jorge Martín |
| T3b Normalización | Sergio Rodríguez |
| T4a Filtrado | Argail |
| T4b Groupby | Daniel (asignada a Ainoha, no entregada) |
| T5 Conclusiones y revisión final | Daniel |

---
# T1 — Carga e inspección
**Autor/a:** Jia Xi Shan

# Carga de datos

In [1]:
import pandas as pd
import numpy as np

# Carga del dataset desde la carpeta Data del repositorio.
# Si no se encuentra (por ejemplo al abrir el notebook en Colab), se descarga
# de la misma fuente original, así que los datos son idénticos.
try:
    df = pd.read_csv("../Data/titanic.csv")
except FileNotFoundError:
    df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

df.shape

(891, 15)

# Mi tarea
Voy a inspeccionar el dataset Titanic para conocer su estructura, dimensiones, tipos de datos y estadísticas descriptivas antes de realizar el análisis.

In [2]:
# Mostramos las primeras filas del dataset.
df.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [3]:
# Mostramos información sobre las columnas y los tipos de datos.
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 15 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   survived     891 non-null    int64  
 1   pclass       891 non-null    int64  
 2   sex          891 non-null    str    
 3   age          714 non-null    float64
 4   sibsp        891 non-null    int64  
 5   parch        891 non-null    int64  
 6   fare         891 non-null    float64
 7   embarked     889 non-null    str    
 8   class        891 non-null    str    
 9   who          891 non-null    str    
 10  adult_male   891 non-null    bool   
 11  deck         203 non-null    str    
 12  embark_town  889 non-null    str    
 13  alive        891 non-null    str    
 14  alone        891 non-null    bool   
dtypes: bool(2), float64(2), int64(4), str(7)
memory usage: 92.4 KB


In [4]:
# Mostramos algunas estadísticas básicas de los datos numéricos.
df.describe()

,survived,pclass,age,sibsp,parch,fare
count,891.000000,891.000000,714.000000,891.000000,891.000000,891.000000
mean,0.383838,2.308642,29.699118,0.523008,0.381594,32.204208
std,0.486592,0.836071,14.526497,1.102743,0.806057,49.693429
min,0.000000,1.000000,0.420000,0.000000,0.000000,0.000000
25%,0.000000,2.000000,20.125000,0.000000,0.000000,7.910400
50%,0.000000,3.000000,28.000000,0.000000,0.000000,14.454200
75%,1.000000,3.000000,38.000000,1.000000,0.000000,31.000000
max,1.000000,3.000000,80.000000,8.000000,6.000000,512.329200


In [5]:
# Mostramos el número de filas y columnas.
df.shape

(891, 15)

## Diccionario de datos

| Columna | Descripción |
|---|---|
| survived | Indica si el pasajero sobrevivió. 0 = No, 1 = Sí |
| pclass | Clase del billete. 1 = Primera, 2 = Segunda, 3 = Tercera |
| sex | Sexo del pasajero |
| age | Edad del pasajero |
| sibsp | Número de hermanos/as o cónyuges a bordo |
| parch | Número de padres/madres o hijos/as a bordo |
| fare | Precio del billete |
| embarked | Código del puerto donde embarcó |
| class | Clase del billete en formato texto |
| who | Categoría del pasajero |
| adult_male | Indica si el pasajero es un hombre adulto |
| deck | Cubierta del barco |
| embark_town | Ciudad donde embarcó |
| alive | Indica si el pasajero sobrevivió |
| alone | Indica si viajaba solo |

## Hallazgo
El dataset tiene 891 pasajeros y 15 columnas. La edad tiene 714 valores y la edad media es de unos 29,70 años. También hay datos numéricos y categóricos, como la edad, el sexo, la clase y la supervivencia.

---
# T2a — Detección y tratamiento de nulos
**Autor/a:** Alejandro

In [6]:
import pandas as pd
import numpy as np

# Carga del dataset desde la carpeta Data del repositorio.
# Si no se encuentra (por ejemplo al abrir el notebook en Colab), se descarga
# de la misma fuente original, así que los datos son idénticos.
try:
    df = pd.read_csv("../Data/titanic.csv")
except FileNotFoundError:
    df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")


# Mostramos las primeras filas
df.head()

,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
0,0,3,male,22.0,1,0,7.2500,S,Third,man,True,NaN,Southampton,no,False
1,1,1,female,38.0,1,0,71.2833,C,First,woman,False,C,Cherbourg,yes,False
2,1,3,female,26.0,0,0,7.9250,S,Third,woman,False,NaN,Southampton,yes,True
3,1,1,female,35.0,1,0,53.1000,S,First,woman,False,C,Southampton,yes,False
4,0,3,male,35.0,0,0,8.0500,S,Third,man,True,NaN,Southampton,no,True


In [7]:
# Mostramos la cantidad de valores nulos por columna
df.isna().sum()

survived         0
pclass           0
sex              0
age            177
sibsp            0
parch            0
fare             0
embarked         2
class            0
who              0
adult_male       0
deck           688
embark_town      2
alive            0
alone            0
dtype: int64

In [8]:
# Hacemos una copia para aplicar el tratamiento
df_limpio = df.copy()

# Imputamos la mediana en la columna age
mediana_edad = df_limpio["age"].median()
df_limpio["age"] = df_limpio["age"].fillna(mediana_edad)

# Imputamos la moda en las columnas embarked y embark_town
moda_embarked = df_limpio["embarked"].mode()[0]
df_limpio["embarked"] = df_limpio["embarked"].fillna(moda_embarked)

moda_town = df_limpio["embark_town"].mode()[0]
df_limpio["embark_town"] = df_limpio["embark_town"].fillna(moda_town)

# Eliminamos la columna deck por exceso de nulos
df_limpio = df_limpio.drop(columns=["deck"])

## Justificación de cada decisión

| Columna | Nulos | Decisión | Por qué |
|---|---|---|---|
| `age` | 177 (19,9%) | Rellenar con la **mediana** (28 años) | Son demasiadas filas para borrarlas: perderíamos 1 de cada 5 pasajeros. Usamos la mediana y no la media (29,7) porque la edad tiene asimetría positiva y algunos valores altos (hasta 80 años) que tiran de la media hacia arriba; la mediana es más robusta frente a esos extremos. |
| `embarked` | 2 (0,2%) | Rellenar con la **moda** (`S`, Southampton) | Es una variable categórica, así que no tiene media ni mediana. Con solo 2 nulos, asignarles el puerto más frecuente apenas altera la distribución. |
| `embark_town` | 2 (0,2%) | Rellenar con la **moda** (Southampton) | Es la misma información que `embarked` en texto; se trata igual para que ambas columnas sigan siendo coherentes. |
| `deck` | 688 (77,2%) | **Eliminar la columna** | Falta en más de 3 de cada 4 pasajeros. Rellenarla sería inventarse la mayoría de los valores, y borrar esas filas dejaría el dataset en 203 pasajeros. |

**Limitación a tener en cuenta:** al imputar la edad, 177 pasajeros pasan a tener exactamente 28 años. Esto concentra artificialmente valores en la franja 18-59, algo que se tiene en cuenta al interpretar los resultados de edad en T5.

In [9]:
# Comprobamos que no queden valores nulos en el dataset
df_limpio.isna().sum()

survived       0
pclass         0
sex            0
age            0
sibsp          0
parch          0
fare           0
embarked       0
class          0
who            0
adult_male     0
embark_town    0
alive          0
alone          0
dtype: int64

## Hallazgo

Se detectaron 869 valores nulos repartidos en `deck` (688), `age` (177) y `embarked`/`embark_town` (2 cada una).

El tratamiento deja el DataFrame con 0 nulos en todas sus variables y conservando íntegras las 891 filas originales.

---
# T2b — Duplicados y calidad de los datos
**Autor/a:** Noel y Daniel

In [10]:
import pandas as pd
import numpy as np

# Carga del dataset desde la carpeta Data del repositorio.
# Si no se encuentra (por ejemplo al abrir el notebook en Colab), se descarga
# de la misma fuente original, así que los datos son idénticos.
try:
    df = pd.read_csv("../Data/titanic.csv")
except FileNotFoundError:
    df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

df.shape

(891, 15)

## 1. Detección de duplicados

In [11]:
# Comprobamos cuántas filas duplicadas hay
# (filas que son copia exacta de otra fila anterior, en todas las columnas)
print("Número de filas duplicadas:", df.duplicated().sum())

# Mostramos las filas que forman parte de un duplicado, incluida la primera aparición,
# para poder compararlas entre sí
mask_dup = df.duplicated(keep=False)
print("Filas implicadas en algún duplicado:", mask_dup.sum())
df[mask_dup].sort_values(["pclass", "sex", "fare"]).head(10)

Número de filas duplicadas: 107
Filas implicadas en algún duplicado: 160


,survived,pclass,sex,age,sibsp,parch,fare,embarked,class,who,adult_male,deck,embark_town,alive,alone
369,1,1,female,24.0,0,0,69.3000,C,First,woman,False,B,Cherbourg,yes,True
641,1,1,female,24.0,0,0,69.3000,C,First,woman,False,B,Cherbourg,yes,True
64,0,1,male,NaN,0,0,27.7208,C,First,man,True,NaN,Cherbourg,no,True
295,0,1,male,NaN,0,0,27.7208,C,First,man,True,NaN,Cherbourg,no,True
458,1,2,female,50.0,0,0,10.5000,S,Second,woman,False,NaN,Southampton,yes,True
526,1,2,female,50.0,0,0,10.5000,S,Second,woman,False,NaN,Southampton,yes,True
443,1,2,female,28.0,0,0,13.0000,S,Second,woman,False,NaN,Southampton,yes,True
635,1,2,female,28.0,0,0,13.0000,S,Second,woman,False,NaN,Southampton,yes,True
53,1,2,female,29.0,1,0,26.0000,S,Second,woman,False,NaN,Southampton,yes,False
133,1,2,female,29.0,1,0,26.0000,S,Second,woman,False,NaN,Southampton,yes,False


## 2. ¿Son duplicados reales?

Este dataset (versión de seaborn) no incluye columnas que identifiquen a cada persona, como el nombre o el número de billete, que sí están en el CSV original de Kaggle. Por eso hay que comprobar si las filas "duplicadas" son errores de carga o pasajeros distintos que simplemente comparten los mismos valores.

In [12]:
# Tamaño de cada grupo de filas idénticas (dropna=False para no descartar
# los grupos que tienen NaN en age o deck)
grupos = (
    df[mask_dup]
    .groupby(list(df.columns), dropna=False)
    .size()
    .sort_values(ascending=False)
)
print("Grupos con exactamente 2 filas idénticas:", (grupos == 2).sum())
print("Grupos con 3 o más filas idénticas:", (grupos >= 3).sum())
print("Tamaño del grupo más grande:", grupos.max())

# ¿Cuántas de las filas duplicadas no tienen edad registrada?
print("Filas duplicadas sin edad (age = NaN):", df.loc[mask_dup, "age"].isna().sum(), "de", mask_dup.sum())

Grupos con exactamente 2 filas idénticas: 35
Grupos con 3 o más filas idénticas: 18
Tamaño del grupo más grande: 13
Filas duplicadas sin edad (age = NaN): 93 de 160


In [13]:
# El grupo más grande: ¿qué perfil de pasajero es?
grupos.head(3)

survived  pclass  sex   age  sibsp  parch  fare    embarked  class  who  adult_male  deck  embark_town  alive  alone
0         3       male  NaN  0      0      7.8958  S         Third  man  True        NaN   Southampton  no     True     13
                                           8.0500  S         Third  man  True        NaN   Southampton  no     True     12
                                           7.7500  Q         Third  man  True        NaN   Queenstown   no     True      8
dtype: int64

## 3. Impacto de eliminar los duplicados

In [14]:
# Comparamos el dataset completo con el que quedaría sin duplicados
tasa_con = df["survived"].mean() * 100
df_sin_dup = df.drop_duplicates()
tasa_sin = df_sin_dup["survived"].mean() * 100

print(f"Filas con todos los datos: {len(df)}")
print(f"Filas si elimináramos los duplicados: {len(df_sin_dup)}")
print(f"Tasa de supervivencia con todas las filas: {tasa_con:.2f}%")
print(f"Tasa de supervivencia sin duplicados: {tasa_sin:.2f}%")

Filas con todos los datos: 891
Filas si elimináramos los duplicados: 784
Tasa de supervivencia con todas las filas: 38.38%
Tasa de supervivencia sin duplicados: 41.20%


## 4. Valores fuera de rango

In [15]:
# Buscamos valores imposibles en las variables numéricas y categóricas
print("Edad mínima / máxima:", df["age"].min(), "/", df["age"].max())
print("Tarifa mínima / máxima:", df["fare"].min(), "/", df["fare"].max())
print("Valores únicos de pclass:", sorted(df["pclass"].unique()))
print("Valores únicos de sex:", df["sex"].unique())

# Pasajeros con tarifa 0: ¿son un error?
tarifa_cero = df[df["fare"] == 0]
print("\nPasajeros con tarifa = 0:", len(tarifa_cero))
tarifa_cero["pclass"].value_counts().sort_index()

Edad mínima / máxima: 0.42 / 80.0
Tarifa mínima / máxima: 0.0 / 512.3292
Valores únicos de pclass: [np.int64(1), np.int64(2), np.int64(3)]
Valores únicos de sex: <StringArray>
['male', 'female']
Length: 2, dtype: str

Pasajeros con tarifa = 0: 15


pclass
1    5
2    6
3    4
Name: count, dtype: int64

## 5. Decisión y justificación

**No se elimina ninguna fila.**

- Sin nombre ni número de billete, `duplicated()` solo compara 15 columnas, casi todas con pocos valores posibles (sexo, clase, puerto, etc.). Es esperable que pasajeros distintos coincidan en todas ellas.
- Los grupos de filas idénticas no son solo parejas: el más grande tiene **13 filas** (hombres de 3ª clase, sin edad registrada, viajando solos y con la misma tarifa barata de Southampton). Es un perfil muy habitual en el Titanic, no un error de carga que duplique a una misma persona.
- **93 de las 160 filas** implicadas en duplicados no tienen edad. La falta de edad elimina una de las pocas columnas que diferenciaría a esos pasajeros, lo que explica buena parte de las coincidencias.
- Eliminarlas quitaría el 12% del dataset (de 891 a 784 filas) y cambiaría la tasa de supervivencia global de 38,38% a 41,20%. Como las filas repetidas son sobre todo hombres de 3ª clase que no sobrevivieron, borrarlas sesgaría el análisis en vez de corregirlo.
- No se han encontrado valores imposibles: edades entre 0,42 y 80 años, tarifas entre 0 y 512,33, y `pclass`/`sex` solo con los valores esperados. Hay 15 pasajeros con tarifa 0, repartidos entre las tres clases; lo más probable es que viajaran con billete gratuito o de cortesía, así que no se tratan como error.

No podemos asegurar al 100% que ninguna fila sea un duplicado real, porque faltan los identificadores. Pero con los datos disponibles, eliminarlas causaría más distorsión de la que evitaría.

## Hallazgo

Hay 107 filas duplicadas (160 filas implicadas en total), pero se decidió **no eliminarlas**. El dataset no tiene columnas identificadoras y los grupos de filas idénticas, de hasta 13 pasajeros, responden a perfiles comunes (sobre todo hombres de 3ª clase viajando solos, 93 de ellos sin edad registrada). Eliminarlas reduciría el dataset un 12% y subiría la tasa de supervivencia global de 38,38% a 41,20%, sesgando el análisis. No se encontraron valores fuera de rango.

---
# T3a — Franja de edad
**Autor/a:** Jorge Martín

In [16]:
import pandas as pd
import numpy as np

# Carga del dataset desde la carpeta Data del repositorio.
# Si no se encuentra (por ejemplo al abrir el notebook en Colab), se descarga
# de la misma fuente original, así que los datos son idénticos.
try:
    df = pd.read_csv("../Data/titanic.csv")
except FileNotFoundError:
    df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

df.shape

(891, 15)

Aplicamos la base común del contrato para trabajar con las edades ya sin nulos (la mediana imputada, ver T2a).

In [17]:
# --- BASE COMÚN (tratamiento de nulos acordado en el contrato, ver T2a) ---
df["age"] = df["age"].fillna(df["age"].median())
df["embarked"] = df["embarked"].fillna(df["embarked"].mode()[0])
df["embark_town"] = df["embark_town"].fillna(df["embark_town"].mode()[0])
df = df.drop(columns="deck")

In [18]:
# Mostramos las primeras 10 edades
df["age"].head(10)

0    22.0
1    38.0
2    26.0
3    35.0
4    35.0
5    28.0
6    54.0
7     2.0
8    27.0
9    14.0
Name: age, dtype: float64

In [19]:
# Definimos las condiciones para cada franja de edad.
# Usamos límites contiguos (< 13, < 18, < 60) para que ninguna edad con decimales
# (por ejemplo 12,5 o 59,5) se quede fuera de todas las franjas.
edad = df["age"]
condiciones = [
    edad < 13,
    (edad >= 13) & (edad < 18),
    (edad >= 18) & (edad < 60),
    edad >= 60,
]

# Definimos el nombre de cada franja
valores = ["0-12", "13-17", "18-59", "60+"]

# Creamos la nueva columna con NumPy, de forma vectorizada (sin bucles).
# El valor por defecto es texto para que toda la columna sea del mismo tipo.
df["franja_edad"] = np.select(condiciones, valores, default="Desconocido")

In [20]:
# Mostramos la edad junto con la nueva franja
df[["age", "franja_edad"]].head(20)

,age,franja_edad
0,22.0,18-59
1,38.0,18-59
2,26.0,18-59
3,35.0,18-59
4,35.0,18-59
5,28.0,18-59
6,54.0,18-59
7,2.0,0-12
8,27.0,18-59
9,14.0,13-17


In [21]:
# Contamos cuántos pasajeros hay en cada franja
df["franja_edad"].value_counts()

franja_edad
18-59    752
0-12      69
13-17     44
60+       26
Name: count, dtype: int64

In [22]:
# Comprobamos las categorías existentes
df["franja_edad"].unique()

<StringArray>
['18-59', '0-12', '13-17', '60+']
Length: 4, dtype: str

## Hallazgo

La columna `franja_edad` agrupa a los 891 pasajeros en cuatro franjas sin dejar ninguno fuera: 69 tienen entre 0 y 12 años, 44 entre 13 y 17, 752 entre 18 y 59, y 26 tienen 60 o más. La franja 18-59 concentra el 84% de los pasajeros, en parte porque incluye a los 177 pasajeros cuya edad se rellenó con la mediana (28 años).

---
# T3b — Normalización y tamaño de familia
**Autor/a:** Sergio Rodríguez

In [23]:
import pandas as pd
import numpy as np

# Carga del dataset desde la carpeta Data del repositorio.
# Si no se encuentra (por ejemplo al abrir el notebook en Colab), se descarga
# de la misma fuente original, así que los datos son idénticos.
try:
    df = pd.read_csv("../Data/titanic.csv")
except FileNotFoundError:
    df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

df.shape

(891, 15)

Aplicamos la base común del contrato (tratamiento de nulos, ver T2a).

In [24]:
# --- BASE COMÚN (tratamiento de nulos acordado en el contrato, ver T2a) ---
df["age"] = df["age"].fillna(df["age"].median())
df["embarked"] = df["embarked"].fillna(df["embarked"].mode()[0])
df["embark_town"] = df["embark_town"].fillna(df["embark_town"].mode()[0])
df = df.drop(columns="deck")

In [25]:
# Tamaño de la familia: hermanos/cónyuges + padres/hijos + el propio pasajero
df["tam_familia"] = df["sibsp"] + df["parch"] + 1

# Normalización min-max de la tarifa: (valor - mínimo) / (máximo - mínimo)
# La operación se aplica a toda la columna a la vez (vectorizada)
fare_min = df["fare"].min()
fare_max = df["fare"].max()
df["tarifa_norm"] = (df["fare"] - fare_min) / (fare_max - fare_min)

df[["tam_familia", "tarifa_norm"]].describe()

,tam_familia,tarifa_norm
count,891.000000,891.000000
mean,1.904602,0.062858
std,1.613459,0.096995
min,1.000000,0.000000
25%,1.000000,0.015440
50%,1.000000,0.028213
75%,2.000000,0.060508
max,11.000000,1.000000


In [26]:
# Comprobamos que la normalización es correcta: el mínimo debe ser 0 y el máximo 1
assert np.isclose(df["tarifa_norm"].min(), 0)
assert np.isclose(df["tarifa_norm"].max(), 1)

# Cuántos pasajeros hay de cada tamaño de familia
print(df["tam_familia"].value_counts().sort_index())

# Relación entre tamaño de familia y tarifa pagada
print(df[["tarifa_norm", "tam_familia"]].corr().round(3))

# Tarifa normalizada media en cada clase
print(df.groupby("pclass")["tarifa_norm"].mean().round(3))

tam_familia
1     537
2     161
3     102
4      29
5      15
6      22
7      12
8       6
11      7
Name: count, dtype: int64
             tarifa_norm  tam_familia
tarifa_norm        1.000        0.217
tam_familia        0.217        1.000
pclass
1    0.164
2    0.040
3    0.027
Name: tarifa_norm, dtype: float64


## Hallazgo

El 60% de los pasajeros (537 de 891) viajaba solo. La tarifa normalizada media en 1ª clase (0,164) fue más de 6 veces la de 3ª clase (0,027), y la relación entre tamaño de familia y tarifa pagada es débil (correlación de 0,217).

---
# T4a — Filtrado
**Autor/a:** Argail

## 1. Carga de datos (todos)

In [27]:
import pandas as pd
import numpy as np

# Carga del dataset desde la carpeta Data del repositorio.
# Si no se encuentra (por ejemplo al abrir el notebook en Colab), se descarga
# de la misma fuente original, así que los datos son idénticos.
try:
    df = pd.read_csv("../Data/titanic.csv")
except FileNotFoundError:
    df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

df.shape

(891, 15)

## 2. Base común (solo P4, P5, P6 y P7)
> **P1, P2 y P3: borrad esta celda y la de código de abajo**, porque necesitáis los datos originales con sus nulos y duplicados.  
> **No la modifiquéis.** Aplica el tratamiento de nulos acordado en el contrato.

In [28]:
# --- BASE COMÚN (no modificar) ---
df["age"] = df["age"].fillna(df["age"].median())
df["embarked"] = df["embarked"].fillna(df["embarked"].mode()[0])
df["embark_town"] = df["embark_town"].fillna(df["embark_town"].mode()[0])
df = df.drop(columns="deck")

## 3. Mi tarea
> Voy a aplicar dos filtros con condiciones distintas y calcular para cada grupo el número de pasajeros y su porcentaje de supervivencia.

In [29]:
# Filtro 1: seleccionamos únicamente a las mujeres.
filtro_mujeres = df[df["sex"] == "female"]

# Filtro 2: seleccionamos únicamente a los pasajeros de primera clase.
filtro_primera_clase = df[df["pclass"] == 1]

# Estadísticas del filtro de mujeres.
num_mujeres = len(filtro_mujeres)
porcentaje_supervivencia_mujeres = filtro_mujeres["survived"].mean() * 100

# Estadísticas del filtro de primera clase.
num_primera_clase = len(filtro_primera_clase)
porcentaje_supervivencia_primera_clase = filtro_primera_clase["survived"].mean() * 100

# Mostramos los resultados de los dos filtros.
print("Mujeres:", num_mujeres)
print("Porcentaje de supervivencia de las mujeres:",
      round(porcentaje_supervivencia_mujeres, 2), "%")

print("Pasajeros de primera clase:", num_primera_clase)
print("Porcentaje de supervivencia de primera clase:",
      round(porcentaje_supervivencia_primera_clase, 2), "%")

Mujeres: 314
Porcentaje de supervivencia de las mujeres: 74.2 %
Pasajeros de primera clase: 216
Porcentaje de supervivencia de primera clase: 62.96 %


### Filtro combinando dos condiciones a la vez

Además de los dos filtros anteriores, combinamos dos condiciones en un mismo filtro con el operador `&` (y), para ver el efecto conjunto de sexo y clase.

In [30]:
# Filtro 3: mujeres Y de primera clase (las dos condiciones a la vez)
filtro_mujeres_primera = df[(df["sex"] == "female") & (df["pclass"] == 1)]

# Filtro 4: hombres adultos (18 años o más) de tercera clase
filtro_hombres_tercera = df[(df["sex"] == "male") & (df["pclass"] == 3) & (df["age"] >= 18)]

print("Mujeres de 1ª clase:", len(filtro_mujeres_primera),
      "- supervivencia:", round(filtro_mujeres_primera["survived"].mean() * 100, 2), "%")
print("Hombres adultos de 3ª clase:", len(filtro_hombres_tercera),
      "- supervivencia:", round(filtro_hombres_tercera["survived"].mean() * 100, 2), "%")

Mujeres de 1ª clase: 94 - supervivencia: 96.81 %
Hombres adultos de 3ª clase: 304 - supervivencia: 12.17 %


## 4. Hallazgo
En el dataset hay 314 mujeres y su porcentaje de supervivencia fue del 74,20%. En primera clase había 216 pasajeros y su porcentaje de supervivencia fue del 62,96%. Al combinar ambas condiciones, las 94 mujeres de 1ª clase sobrevivieron en un 96,81%, frente al 12,17% de los 304 hombres adultos de 3ª clase.

---
# T4b — Estadísticas por categoría (groupby)
**Autor/a:** Daniel (asignada a Ainoha, no entregada)

In [31]:
import pandas as pd
import numpy as np

# Carga del dataset desde la carpeta Data del repositorio.
# Si no se encuentra (por ejemplo al abrir el notebook en Colab), se descarga
# de la misma fuente original, así que los datos son idénticos.
try:
    df = pd.read_csv("../Data/titanic.csv")
except FileNotFoundError:
    df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

df.shape

(891, 15)

Aplicamos la base común del contrato (tratamiento de nulos, ver T2a).

In [32]:
# --- BASE COMÚN (tratamiento de nulos acordado en el contrato, ver T2a) ---
df["age"] = df["age"].fillna(df["age"].median())
df["embarked"] = df["embarked"].fillna(df["embarked"].mode()[0])
df["embark_town"] = df["embark_town"].fillna(df["embark_town"].mode()[0])
df = df.drop(columns="deck")

## 1. Tasa de supervivencia por sexo

In [33]:
# Media de survived (0/1) por sexo = proporción de supervivientes, en %
tasa_sexo = (df.groupby("sex")["survived"].mean() * 100).round(2)
tasa_sexo

sex
female    74.20
male      18.89
Name: survived, dtype: float64

## 2. Tasa de supervivencia por clase

In [34]:
tasa_clase = (df.groupby("pclass")["survived"].mean() * 100).round(2)
tasa_clase

pclass
1    62.96
2    47.28
3    24.24
Name: survived, dtype: float64

## 3. Tasa de supervivencia por sexo y clase combinados
Agrupamos por las dos variables a la vez para ver si combinarlas cambia el panorama respecto a mirarlas por separado. Con `unstack()` lo mostramos como tabla (filas: sexo, columnas: clase).

In [35]:
tasa_sexo_clase = (df.groupby(["sex", "pclass"])["survived"].mean() * 100).round(2)
tasa_sexo_clase.unstack()

pclass,1,2,3
sex,,,
female,96.81,92.11,50.00
male,36.89,15.74,13.54


## 4. Tarifa y edad media por clase

In [36]:
# Varias estadísticas a la vez: media de tarifa y de edad por clase
df.groupby("pclass")[["fare", "age"]].mean().round(2)

,fare,age
pclass,,
1,84.15,36.81
2,20.66,29.77
3,13.68,25.93


## Hallazgo

Por separado, el sexo marca más diferencia que la clase: sobrevivió el 74,20% de las mujeres frente al 18,89% de los hombres, mientras que por clase el rango va del 62,96% (1ª) al 24,24% (3ª). Al cruzar ambas variables aparece algo que no se ve por separado: entre las mujeres, la clase casi no importa en 1ª y 2ª (96,81% y 92,11%) y la caída llega en 3ª (50,00%); entre los hombres, la supervivencia es baja en todas las clases (36,89%, 15,74% y 13,54%). En puntos porcentuales, pasar de 1ª a 3ª clase supuso una caída de 46,8 puntos para las mujeres y de 23,4 para los hombres. Además, la tarifa media en 1ª clase (84,15) es más de 6 veces la de 3ª (13,68), lo que confirma que clase y poder adquisitivo van de la mano.

---
*Nota: esta tarea estaba asignada a Ainoha y no se entregó, así que la completó Daniel como jefe de grupo para poder cerrar la entrega.*

---
# T5 — Preguntas guía y conclusiones finales
**Autor/a:** Daniel

In [37]:
import pandas as pd
import numpy as np

# Carga del dataset desde la carpeta Data del repositorio.
# Si no se encuentra (por ejemplo al abrir el notebook en Colab), se descarga
# de la misma fuente original, así que los datos son idénticos.
try:
    df = pd.read_csv("../Data/titanic.csv")
except FileNotFoundError:
    df = pd.read_csv("https://raw.githubusercontent.com/mwaskom/seaborn-data/master/titanic.csv")

df.shape

(891, 15)

Aplicamos la base común del contrato (nulos, ver T2a) y recreamos la franja de edad de T3a. Guardamos también una copia de los datos originales para comprobar el efecto de la imputación de edad.

In [38]:
df_original = df.copy()  # datos sin tratar, para la pregunta 5

# --- BASE COMÚN (tratamiento de nulos acordado en el contrato, ver T2a) ---
df["age"] = df["age"].fillna(df["age"].median())
df["embarked"] = df["embarked"].fillna(df["embarked"].mode()[0])
df["embark_town"] = df["embark_town"].fillna(df["embark_town"].mode()[0])
df = df.drop(columns="deck")

# Recreamos la franja de edad igual que en T3a
edad = df["age"]
condiciones = [edad < 13, (edad >= 13) & (edad < 18), (edad >= 18) & (edad < 60), edad >= 60]
valores = ["0-12", "13-17", "18-59", "60+"]
df["franja_edad"] = np.select(condiciones, valores, default="Desconocido")

## Pregunta 1: ¿Qué porcentaje de pasajeros sobrevivió en total?

In [39]:
print("Supervivientes:", df["survived"].sum(), "de", len(df))
round(df["survived"].mean() * 100, 2)

Supervivientes: 342 de 891


np.float64(38.38)

**Respuesta:** sobrevivió el 38,38% de los pasajeros (342 de 891).

## Pregunta 2: ¿Cómo varía la tasa de supervivencia según el sexo?

In [40]:
(df.groupby("sex")["survived"].mean() * 100).round(2)

sex
female    74.20
male      18.89
Name: survived, dtype: float64

**Respuesta:** sobrevivió el 74,20% de las mujeres frente al 18,89% de los hombres. Es la variable individual que más diferencia marca en todo el dataset.

## Pregunta 3: ¿Y según la clase del billete (1ª, 2ª, 3ª)?

In [41]:
(df.groupby("pclass")["survived"].mean() * 100).round(2)

pclass
1    62.96
2    47.28
3    24.24
Name: survived, dtype: float64

**Respuesta:** 62,96% en 1ª clase, 47,28% en 2ª y 24,24% en 3ª. La supervivencia baja de forma constante a medida que baja la clase.

## Pregunta 4: ¿Existe relación entre la edad y la probabilidad de supervivencia?

In [42]:
print("Correlación edad-supervivencia (con edades imputadas):", round(df["age"].corr(df["survived"]), 3))
print("Correlación edad-supervivencia (solo edades reales):", round(df_original["age"].corr(df_original["survived"]), 3))

Correlación edad-supervivencia (con edades imputadas): -0.065
Correlación edad-supervivencia (solo edades reales): -0.077


**Respuesta:** la correlación lineal entre edad y supervivencia es débil y negativa (-0,065; -0,077 usando solo las edades reales), casi nula. Pero eso no significa que la edad no influya: agrupando por franjas (pregunta 5) sí aparece una relación, concentrada en los pasajeros más jóvenes y los más mayores.

## Pregunta 5: ¿Qué franja de edad tiene mayor proporción de supervivientes?

In [43]:
# Tasa de supervivencia (media de survived) y número de pasajeros por franja
tabla_edad = df.groupby("franja_edad")["survived"].agg(["mean", "size"])
tabla_edad["mean"] = (tabla_edad["mean"] * 100).round(2)
tabla_edad.columns = ["tasa_supervivencia_%", "num_pasajeros"]
tabla_edad

,tasa_supervivencia_%,num_pasajeros
franja_edad,,
0-12,57.97,69
13-17,47.73,44
18-59,36.44,752
60+,26.92,26


In [44]:
# Comprobación: la misma tabla usando solo los pasajeros con edad real (sin imputar)
reales = df_original.dropna(subset=["age"]).copy()
edad_r = reales["age"]
reales["franja_edad"] = np.select(
    [edad_r < 13, (edad_r >= 13) & (edad_r < 18), (edad_r >= 18) & (edad_r < 60), edad_r >= 60],
    valores, default="Desconocido")
(reales.groupby("franja_edad")["survived"].mean() * 100).round(2)

franja_edad
0-12     57.97
13-17    47.73
18-59    38.61
60+      26.92
Name: survived, dtype: float64

**Respuesta:** la franja 0-12 años tiene la mayor proporción de supervivientes (57,97%), por delante de 13-17 (47,73%), 18-59 (36,44%) y 60+ (26,92%). La supervivencia baja de una franja a la siguiente, lo que es coherente con la prioridad a "mujeres y niños" en los botes. Aun así, la correlación de la pregunta 4 es casi nula porque la diferencia se concentra en los extremos (niños y mayores de 60), que son pocos pasajeros (95 de 891), mientras que la gran mayoría (84%) está en la franja 18-59.

La imputación de la edad solo afecta a la franja 18-59, que es donde caen los 177 pasajeros sin edad: con ellos baja del 38,61% al 36,44%, porque esos pasajeros sobrevivieron menos (29,38%). Las demás franjas no cambian y el orden entre franjas es el mismo.

## Pregunta 6 (opcional): ¿La combinación de sexo y clase cambia mucho la tasa de supervivencia respecto a mirar cada variable por separado?

In [45]:
(df.groupby(["sex", "pclass"])["survived"].mean() * 100).round(2).unstack()

pclass,1,2,3
sex,,,
female,96.81,92.11,50.00
male,36.89,15.74,13.54


**Respuesta:** sí. Por separado, podría parecer que sexo y clase suman sus efectos de forma independiente. Al cruzarlos se ve que, entre las mujeres, la clase apenas cambia la supervivencia en 1ª y 2ª (96,81% y 92,11%) y solo cae con fuerza en 3ª (50,00%). Entre los hombres, la supervivencia es baja en las tres clases (36,89%, 15,74% y 13,54%). Viajar en 1ª clase protegía mucho más a una mujer (+46,8 puntos respecto a 3ª) que a un hombre (+23,4 puntos).

## Conclusiones finales

1. **El sexo es la variable que más determina la supervivencia.** Sobrevivió el 74,20% de las mujeres frente al 18,89% de los hombres, 55 puntos de diferencia, más que cualquier otra variable analizada.

2. **La clase del billete también influye, y de forma escalonada.** La supervivencia baja del 62,96% en 1ª clase al 47,28% en 2ª y al 24,24% en 3ª. La tarifa media en 1ª es más de 6 veces la de 3ª, así que clase y poder adquisitivo van juntos.

3. **Sexo y clase se combinan, no actúan por separado.** Casi todas las mujeres de 1ª y 2ª clase sobrevivieron (más del 92%), pero solo la mitad de las de 3ª. Entre los hombres la supervivencia fue baja en todas las clases. Mirar las dos variables juntas cuenta una historia distinta a mirarlas por separado.

4. **La edad influye, sobre todo en los extremos.** La correlación entre edad y supervivencia es casi nula (-0,065), pero al agrupar por franjas los niños de 0-12 años tienen la mayor tasa de supervivencia (57,97%) y los mayores de 60 la menor (26,92%). Como son pocos pasajeros, el efecto casi no se nota en la correlación global.

5. **Los datos tienen limitaciones.** Cerca del 20% de las edades son estimaciones (mediana imputada), lo que baja algo la tasa de la franja 18-59. Además, faltan columnas identificadoras (nombre, billete), así que no se puede confirmar si las 107 filas repetidas son pasajeros distintos; con los datos disponibles lo más probable es que sí, y por eso no se eliminaron (ver T2b).